# Introduction

In this homework assignment, I compared two langauge model using two well known children stories. The goal of this was to see how well these models understood the stories and their capability to recreate them. Some metrics that we will look at to understand the language model's performance are loss, perplexity and accuracy.

# Text Chosen

In this assignment, I have chosen two classic children short stories as our example texts.

1. The Tortoise and the Hare
2. The Three Little Pigs

By evaluating language models on these well known stories, we can try to assess the model's ability to understand and generate coherent text across different narritive styles.

# Table that shows the accuracy, loss, and perplexity for each modle on each text

In [1]:
from transformers import pipeline, AutoTokenizer, AutoModelForCausalLM

pipe_phi = pipeline("text-generation", model="microsoft/phi-2", trust_remote_code=True)
pipe_gpt2 = pipeline("text-generation", model="openai-community/gpt2")

tokenizer_phi = AutoTokenizer.from_pretrained("microsoft/phi-2", trust_remote_code=True)
model_phi = AutoModelForCausalLM.from_pretrained("microsoft/phi-2", trust_remote_code=True)

tokenizer_gpt2 = AutoTokenizer.from_pretrained("openai-community/gpt2")
model_gpt2 = AutoModelForCausalLM.from_pretrained("openai-community/gpt2")

2024-03-29 02:12:23.650745: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-03-29 02:12:23.650858: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-03-29 02:12:23.792563: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered


config.json:   0%|          | 0.00/863 [00:00<?, ?B/s]

configuration_phi.py:   0%|          | 0.00/9.26k [00:00<?, ?B/s]

A new version of the following files was downloaded from https://huggingface.co/microsoft/phi-2:
- configuration_phi.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.


modeling_phi.py:   0%|          | 0.00/62.7k [00:00<?, ?B/s]

A new version of the following files was downloaded from https://huggingface.co/microsoft/phi-2:
- modeling_phi.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.


model.safetensors.index.json:   0%|          | 0.00/35.7k [00:00<?, ?B/s]

model-00001-of-00002.safetensors:   0%|          | 0.00/5.00G [00:00<?, ?B/s]

model-00002-of-00002.safetensors:   0%|          | 0.00/564M [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/7.34k [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/798k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/2.11M [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/1.08k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/99.0 [00:00<?, ?B/s]

Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.


config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.


Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

## Prefix Chosen

The prefix that I chose for the first story, The Tortoise and the Hare, was the first sentence of the short story, "A Hare was making fun of the Tortoise one day for being so slow".<br> The prefix that I chose for the second story, The Three Little Pigs, was "Once upon a time there were three little pigs."

## Loss

In [3]:
from transformers import GPT2Tokenizer

tokenizer = GPT2Tokenizer.from_pretrained("gpt2", pad_token="[PAD]")

story1_text = """
A Hare was making fun of the Tortoise one day for being so slow.
"Do you ever get anywhere?" he asked with a mocking laugh.
"Yes," replied the Tortoise, "and I get there sooner than you think. I'll run you a race and prove it."
The Hare was much amused at the idea of running a race with the Tortoise, but for the fun of the thing he agreed. So the Fox, who had consented to act as judge, marked the distance and started the runners off.
The Hare was soon far out of sight, and to make the Tortoise feel very deeply how ridiculous it was for him to try a race with a Hare, he lay down beside the course to take a nap until the Tortoise should catch up.
The Tortoise meanwhile kept going slowly but steadily, and, after a time, passed the place where the Hare was sleeping. But the Hare slept on very peacefully; and when at last he did wake up, the Tortoise was near the goal. The Hare now ran his swiftest, but he could not overtake the Tortoise in time.
"""

story2_text = """Once upon a time there were three little pigs. One pig
built a house of straw while the second pig built his
house with sticks. They built their houses very quickly
and then sang and danced all day because they were
lazy. The third little pig worked hard all day and built
his house with bricks.
A big bad wolf saw the two little pigs while they danced
and played and thought, “What juicy tender meals they
will make!” He chased the two pigs and they ran and
hid in their houses. The big bad wolf went to the first
house and huffed and puffed and blew the house down in minutes. The frightened little
pig ran to the second pig’s house that was made of sticks. The big bad wolf now came to
this house and huffed and puffed and blew the house down in hardly any time. Now, the
two little pigs were terrified and ran to the third pig’s house that was made of bricks.
The big bad wolf tried to huff and puff and blow the house down, but he could not. He
kept trying for hours but the house was very strong and the little pigs were safe inside.
He tried to enter through the chimney but the third little pig boiled a big pot of water
and kept it below the chimney. The wolf fell into it and died.
The two little pigs now felt sorry for having been so lazy. They too built their houses with
bricks and lived happily ever after.
"""

story1_input_ids = tokenizer_phi(story1_text, return_tensors="pt").input_ids
story2_input_ids = tokenizer_phi(story2_text, return_tensors="pt").input_ids

"""
story1_tokens = tokenizer(story1_text, return_tensors="pt", padding=True)
story2_tokens = tokenizer(story2_text, return_tensors="pt", padding=True)

story1_prompt = "A Hare was making fun of the Tortoise one day for being so slow."
story2_prompt = "Once upon a time there were three little pigs."

story1_prompt = story1_prompt.ljust(story1_tokens.input_ids.size(1))
story2_prompt = story2_prompt.ljust(story2_tokens.input_ids.size(1))

story1_input_ids = tokenizer.encode(story1_prompt, return_tensors="pt")[:, :-1]
story2_input_ids = tokenizer.encode(story2_prompt, return_tensors="pt")[:, :-1]
"""

'\nstory1_tokens = tokenizer(story1_text, return_tensors="pt", padding=True)\nstory2_tokens = tokenizer(story2_text, return_tensors="pt", padding=True)\n\nstory1_prompt = "A Hare was making fun of the Tortoise one day for being so slow."\nstory2_prompt = "Once upon a time there were three little pigs."\n\nstory1_prompt = story1_prompt.ljust(story1_tokens.input_ids.size(1))\nstory2_prompt = story2_prompt.ljust(story2_tokens.input_ids.size(1))\n\nstory1_input_ids = tokenizer.encode(story1_prompt, return_tensors="pt")[:, :-1]\nstory2_input_ids = tokenizer.encode(story2_prompt, return_tensors="pt")[:, :-1]\n'

## Accuracy

In [13]:
print("\nOriginal 01(The Tortoise and the Hare):")
print(story1_text)
print("\nOriginal 02(The Three Little Pigs):")
print(story2_text)
print()


Original 01(The Tortoise and the Hare):

A Hare was making fun of the Tortoise one day for being so slow.
"Do you ever get anywhere?" he asked with a mocking laugh.
"Yes," replied the Tortoise, "and I get there sooner than you think. I'll run you a race and prove it."
The Hare was much amused at the idea of running a race with the Tortoise, but for the fun of the thing he agreed. So the Fox, who had consented to act as judge, marked the distance and started the runners off.
The Hare was soon far out of sight, and to make the Tortoise feel very deeply how ridiculous it was for him to try a race with a Hare, he lay down beside the course to take a nap until the Tortoise should catch up.
The Tortoise meanwhile kept going slowly but steadily, and, after a time, passed the place where the Hare was sleeping. But the Hare slept on very peacefully; and when at last he did wake up, the Tortoise was near the goal. The Hare now ran his swiftest, but he could not overtake the Tortoise in time.




In [10]:
import torch

num_extra_sentences = 3

print("Generated sentences for story 1 with phi-2 model:")
for _ in range(num_extra_sentences):
    with torch.no_grad():
        outputs = model_phi.generate(
            story1_input_ids,
            max_length=300,  # Increase max_length to prevent truncation
            num_return_sequences=1,
            pad_token_id=tokenizer_phi.eos_token_id,
            attention_mask=story1_input_ids.new_ones(story1_input_ids.shape)
        )
    generated_text = tokenizer_phi.decode(outputs[0], skip_special_tokens=True)
    print(generated_text)

print("\nGenerated sentences for story 2 with phi-2 model:")
for _ in range(num_extra_sentences):
    with torch.no_grad():
        outputs = model_phi.generate(
            story2_input_ids,
            max_length=400,  # Increase max_length to prevent truncation
            num_return_sequences=1,
            pad_token_id=tokenizer_phi.eos_token_id,
            attention_mask=story2_input_ids.new_ones(story2_input_ids.shape)
        )
    generated_text = tokenizer_phi.decode(outputs[0], skip_special_tokens=True)
    print(generated_text)

print("\nGenerated sentences for story 1 with GPT-2 model:")
for _ in range(num_extra_sentences):
    with torch.no_grad():
        outputs = model_gpt2.generate(
            story1_input_ids,
            max_length=300,  # Increase max_length to prevent truncation
            num_return_sequences=1,
            pad_token_id=tokenizer_gpt2.eos_token_id,
            attention_mask=story1_input_ids.new_ones(story1_input_ids.shape)
        )
    generated_text = tokenizer_gpt2.decode(outputs[0], skip_special_tokens=True)
    print(generated_text)

print("\nGenerated sentences for story 2 with GPT-2 model:")
for _ in range(num_extra_sentences):
    with torch.no_grad():
        outputs = model_gpt2.generate(
            story2_input_ids,
            max_length=400,  # Increase max_length to prevent truncation
            num_return_sequences=1,
            pad_token_id=tokenizer_gpt2.eos_token_id,
            attention_mask=story2_input_ids.new_ones(story2_input_ids.shape)
        )
    generated_text = tokenizer_gpt2.decode(outputs[0], skip_special_tokens=True)
    print(generated_text)


Generated sentences for story 1 with phi-2 model:

A Hare was making fun of the Tortoise one day for being so slow.
"Do you ever get anywhere?" he asked with a mocking laugh.
"Yes," replied the Tortoise, "and I get there sooner than you think. I'll run you a race and prove it."
The Hare was much amused at the idea of running a race with the Tortoise, but for the fun of the thing he agreed. So the Fox, who had consented to act as judge, marked the distance and started the runners off.
The Hare was soon far out of sight, and to make the Tortoise feel very deeply how ridiculous it was for him to try a race with a Hare, he lay down beside the course to take a nap until the Tortoise should catch up.
The Tortoise meanwhile kept going slowly but steadily, and, after a time, passed the place where the Hare was sleeping. But the Hare slept on very peacefully; and when at last he did wake up, the Tortoise was near the goal. The Hare now ran his swiftest, but he could not overtake the Tortoise in

In [15]:
with torch.no_grad():
    outputs1_phi = model_phi(story1_input_ids, labels=story1_input_ids)
    loss1_phi = outputs1_phi.loss

with torch.no_grad():
    outputs2_phi = model_phi(story2_input_ids, labels=story2_input_ids)
    loss2_phi = outputs2_phi.loss

with torch.no_grad():
    outputs1_gpt2 = model_gpt2(story1_input_ids, labels=story1_input_ids)
    loss1_gpt2 = outputs1_gpt2.loss

with torch.no_grad():
    outputs2_gpt2 = model_gpt2(story2_input_ids, labels=story2_input_ids)
    loss2_gpt2 = outputs2_gpt2.loss

print("Loss for story 1 with phi-2 model:", loss1_phi.item())
print("Loss for story 2 with phi-2 model:", loss2_phi.item())
print("Loss for story 1 with GPT-2 model:", loss1_gpt2.item())
print("Loss for story 2 with GPT-2 model:", loss2_gpt2.item())

Loss for story 1 with phi-2 model: 0.8170398473739624
Loss for story 2 with phi-2 model: 1.4349267482757568
Loss for story 1 with GPT-2 model: 3.0347375869750977
Loss for story 2 with GPT-2 model: 2.970585584640503


## Perplexity

In [16]:
import math

perplexity1_phi = math.exp(loss1_phi.item())
perplexity2_phi = math.exp(loss2_phi.item())

perplexity1_gpt2 = math.exp(loss1_gpt2.item())
perplexity2_gpt2 = math.exp(loss2_gpt2.item())

print("Perplexity for story 1 with phi-2 model:", perplexity1_phi)
print("Perplexity for story 2 with phi-2 model:", perplexity2_phi)
print("Perplexity for story 1 with GPT-2 model:", perplexity1_gpt2)
print("Perplexity for story 2 with GPT-2 model:", perplexity2_gpt2)


Perplexity for story 1 with phi-2 model: 2.2637887492991386
Perplexity for story 2 with phi-2 model: 4.199337388816805
Perplexity for story 1 with GPT-2 model: 20.795520168065387
Perplexity for story 2 with GPT-2 model: 19.50333710739385


| Model | Text | Accuracy | Loss | Perplexity |
|--------|------|---------|------|------------|
| Model 1(Phi-2) | story 1 | 0.00 | 0.81703984 | 2.2637887 |
| Model 1(Phi-2) | story 2 | 0.00 | .3.03473758 | 4.1993373 |
| Model 2(openai-community/gpt2) | story 1 | 0.00 | 3.03473758 | 20.79552 |
| Model 2(openai-community/gpt2) | story 2 | 0.00 | 2.97058558 | 19.503337 | 

# Interpretation of the results

After reviewing the measurements of both models, it appeared that Microsoft's Phi-2 model generally performed better than GPT-2 in terms of grasping the stories. Microsoft's Phi-2 showed less confusion and stayed more faithful to the original text. However, it was evident that both models struggled to reproduce the stories accurately.

# Conclussion

In Conclusion, neither the GPT2 or Phi-2 model were successful in creating new senteces for the stories. But the Phi-2 model did seem to perform better than GPT-2 when we started to look at loss and perplexity. This means that the Phi-2 model would understand the structure of the story and made a couple less errors compared to the GPT2 model